### PR TITLE
test(lint): harden CtrlProxy getInstance restore guard against alias-shadowing, var-hoisting, and spy-restore false results (#7065)

### LINKED ISSUE #7065: getInstance leak-guard: resolve symbols by lexical binding, fix var function-scope, and let direct restoration clear spy installs
Follow-up from three round-3 review threads on [#7062](https://github.com/kaeawc/auto-mobile/pull/7062) (issue #7052). #7062 already hardened the `test/lint/ctrlProxyGetInstanceRestore.test.ts` guard (import aliases, per-symbol capture matching, direct-vs-spy install split) and is a strict improvement over the guard in main. These three residual edge cases are theoretical — no test file in the current tree triggers them (the guard passes on the real tree) — so per the anti-thrash round limit they are deferred rather than spun into a 4th round on the same PR. Two of them (threads 2 and 3) are latent FALSE POSITIVES that would wrongly flag a clean future test, so they are worth prioritizing.

Verbatim reviewer threads (chatgpt-codex-connector), `test/lint/ctrlProxyGetInstanceRestore.test.ts`:

1. @:118 — **Resolve import aliases by lexical binding**
> When an imported alias is shadowed by a parameter or local, this file-wide name lookup still treats the shadowing identifier as the CtrlProxy import. For example, after an unrestored `Client.getInstance = mock(...)`, a function parameter named `Client` can capture and reassign its unrelated `getInstance`; the analyzer records that as the imported singleton's restore and reports no leak. Resolve the actual lexical/compiler symbol rather than identifier text.

2. @:258 — **Preserve function scope for var declarations** (false positive)
> When an original value is captured with `var` inside a nested block and restored outside that block, `declarationKey` records the binding under the nearest block even though JavaScript scopes `var` to the containing function. The later reference cannot find that key, so the genuine restore is classified as another install and a clean test fails the guard. Choose the containing function/source scope for `var`, while retaining lexical scope for `let` and `const`.

3. @:393 — **Let direct restoration clear spy installs** (false positive)
> When a spy handle is assigned to `getInstance` and teardown assigns a previously captured original value back, the singleton seam is restored correctly, but this branch ignores `facts.restores` and requires `.mockRestore()`. Consequently, a valid sequence such as capturing the original, assigning the spy handle, and assigning the original back is rejected as a leak. Accept either a matching captured-original restoration or the matching spy handle's `.mockRestore()`.

Suggested fixes are inline in each thread; all are contained to the guard test file and want self-tests (including the two false-positive cases as clean-should-pass fixtures).

### PR DESCRIPTION
**Pull request by kaeawc**
Closes the shadowing / var-scope / spy-restore precision gaps in the CtrlProxy `getInstance` leak guard (`test/lint/ctrlProxyGetInstanceRestore.test.ts`), the three follow-ups filed from [#7062](https://github.com/kaeawc/auto-mobile/pull/7062):

1. Symbol resolution is now scope-aware: an identifier resolves to a CtrlProxy import alias only when no shadowing local/parameter intervenes, closing a false negative where a shadowing name could mask an unrestored install.
2. `var` captures are scoped to their containing function (matching JS hoisting) instead of the nearest block, fixing a false positive where a clean capture/restore across a nested block was misclassified as a leak.
3. A direct restore of a captured original now also acquits a spy-handle install (not only `.mockRestore()` on that spy), fixing a false positive on a legitimate capture -> spy-install -> direct-restore sequence.

Three new self-tests (`HOLE 1/2/3`) reproduce each bug. The real-tree scan (`test/daemon`, `test/server`, `test/lint`) stays green and no inventoried installer files were touched.

Closes [#7065](https://github.com/kaeawc/auto-mobile/issues/7065)

### CODE REVIEW

File: test/lint/ctrlProxyGetInstanceRestore.test.ts [hunk 1]
```diff
@@ -390,7 +426,7 @@ describe("CtrlProxy getInstance mocks are restored in-file (issue #7052)", () =>
       }
     }
     for (const symbol of facts.spyInstalls) {
-      if (!facts.restoredBySpy.has(symbol)) {
+      if (!facts.restoredBySpy.has(symbol) && !facts.restores.has(symbol)) {
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply test/lint/ctrlProxyGetInstanceRestore.test.ts]: Fixed in 619893f0f. Install/restore events now carry AST source position (with `afterEach`/`afterAll` callback bodies treated as running last), and `leaksOf` requires the acquitting restore to strictly follow the spy install, so a spy left installed after an earlier direct restore is flagged as a leak while the legitimate capture -> spy-install -> restore sequence still passes.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: test/lint/ctrlProxyGetInstanceRestore.test.ts [hunk 2]
```diff
@@ -267,6 +281,28 @@ describe("CtrlProxy getInstance mocks are restored in-file (issue #7052)", () =>
...
 
+    for (const [alias, symbol] of symbolAliases) {
+      const binding = declarationKey(alias);
+      declaredBindings.add(binding);
+      ctrlProxyBindings.set(binding, symbol);
+    }
+
+    const predeclareWalk = (node: ts.Node): void => {
+      if ((ts.isVariableDeclaration(node) || ts.isParameter(node)) && ts.isIdentifier(node.name)) {
+        declaredBindings.add(declarationKey(node.name));
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply test/lint/ctrlProxyGetInstanceRestore.test.ts:363]: Fixed in 619893f0f. Loop statements (`for`/`for-in`/`for-of`/`while`) now get their own entry in the scope chain, so a `let`/`const` loop variable no longer collides with a same-named import alias's enclosing-block scope, and a real pre-loop `getInstance` install is detected again.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: test/lint/ctrlProxyGetInstanceRestore.test.ts [hunk 3]
```diff
@@ -228,18 +256,32 @@ describe("CtrlProxy getInstance mocks are restored in-file (issue #7052)", () =>
...
+          ts.isSourceFile(current) ||
+          // `let` and `const` declared by a loop initializer are scoped to that
+          // loop, not its enclosing block. Without the loop in this chain,
+          // predeclaration lets a loop-local CtrlProxy-alias look like it shadows
+          // the import throughout the enclosing function, including before the
+          // loop begins.
+          ts.isForStatement(current) ||
+          ts.isForInStatement(current) ||
+          ts.isForOfStatement(current) ||
+          ts.isWhileStatement(current)
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply test/lint/ctrlProxyGetInstanceRestore.test.ts]: Fixed in b2ff3bc50. Catch clauses and switch `CaseBlock`s are now their own lexical scope (via a single `introducesLexicalScope` predicate walking the real parent chain), so a catch-parameter or case-local `Client` alias can no longer shadow an earlier real import-alias install and hide the leak.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: test/lint/ctrlProxyGetInstanceRestore.test.ts [hunk 4]
```diff
@@ -253,8 +295,22 @@ describe("CtrlProxy getInstance mocks are restored in-file (issue #7052)", () =>
...
+    const isVarDeclaration = (name: ts.Identifier): boolean => {
+      const declaration = name.parent;
+      if (
+        !ts.isVariableDeclaration(declaration) ||
+        !ts.isVariableDeclarationList(declaration.parent)
+      ) {
+        return false;
+      }
+      const flags = declaration.parent.flags;
+      return (flags & (ts.NodeFlags.Let | ts.NodeFlags.Const)) === 0;
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply test/lint/ctrlProxyGetInstanceRestore.test.ts]: Fixed in b2ff3bc50. `using`/`await using` declarations are now classified as block-scoped (excluded from the function-scoped check alongside `let`/`const`), so a nested `using Client = resource` no longer shadows an outer real install across the whole function.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: test/lint/ctrlProxyGetInstanceRestore.test.ts [hunk 5]
```diff
@@ -107,29 +115,49 @@ describe("CtrlProxy getInstance mocks are restored in-file (issue #7052)", () =>
...
+    for (let current = node.parent; current; current = current.parent) {
+      if (!ts.isArrowFunction(current) && !ts.isFunctionExpression(current)) {
+        continue;
+      }
+      const call = current.parent;
+      if (!ts.isCallExpression(call) || !call.arguments.includes(current)) {
+        continue;
+      }
+      const callee = unwrap(call.expression);
+      return ts.isIdentifier(callee) && (callee.text === "afterEach" || callee.text === "afterAll");
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply test/lint/ctrlProxyGetInstanceRestore.test.ts]: Fixed in b2ff3bc50. The teardown-ordering walk now climbs past any number of intervening non-hook call arguments (e.g. `helper(() => {...})`) until it reaches a genuine `afterEach`/`afterAll` argument or exhausts ancestors, so a restore nested inside a synchronous helper callback under `afterEach` is recognized as teardown-ordered instead of a false-positive leak.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: test/lint/ctrlProxyGetInstanceRestore.test.ts [hunk 6]
```diff
@@ -253,8 +305,35 @@ describe("CtrlProxy getInstance mocks are restored in-file (issue #7052)", () =>
       }
       return `${id}:${name}`;
     };
+    const isFunctionScopedDeclaration = (name: ts.Identifier): boolean => {
+      const declaration = name.parent;
+      if (ts.isFunctionDeclaration(declaration)) {
+        return true;
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply test/lint/ctrlProxyGetInstanceRestore.test.ts:311]: Deferred to follow-up https://github.com/kaeawc/auto-mobile/issues/7079. This is a theoretical false negative (no test in the tree triggers it; CI green) and the shadow-aware scope model has taken many incremental rounds; per the round-limit policy this and the named-teardown case are batched into a single deliberate pass rather than another round here. This PR already strictly improves guard detection over the prior version.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_

File: test/lint/ctrlProxyGetInstanceRestore.test.ts [hunk 7]
```diff
@@ -107,29 +115,56 @@ describe("CtrlProxy getInstance mocks are restored in-file (issue #7052)", () =>
...
+   */
+  function runsInTeardownCallback(node: ts.Node): boolean {
+    for (let current = node.parent; current;) {
+      if (!ts.isFunctionLike(current)) {
+        current = current.parent;
+        continue;
+      }
+      const call = current.parent;
+      if (!ts.isCallExpression(call) || !call.arguments.some((argument) => argument === current)) {
+        return false;
```
[the remark this answers is not in the document]
Comment (kaeawc) [reply test/lint/ctrlProxyGetInstanceRestore.test.ts:132]: Deferred to follow-up https://github.com/kaeawc/auto-mobile/issues/7079, flagged there as the priority item since it's a latent false positive on the common `afterEach(restoreSingleton)` form (no real test currently hits it, so CI is green, but it would block a correctly-written future test). Batching it with the block-scoped-function case into one deliberate scope-model pass per the round-limit policy.

_🤖 Addressed by [Claude Code](https://claude.com/claude-code)_
